### ISSUE TITLE
WS2 -  crashes application

### ISSUE THREAD

**Issue by Steejayy** [bitfinexcom/bitfinex-api-node#396]
Hello Guys, 

I'm using your implementation of websocoket to trade on Ethfinex. 
Sometimes, after inactivity of 10 mins, 20 mins or 30 minutes, my application seems to crach, I suspect it is linked to the websocket connection. 

Idon't understand why, in fact I registered to the on error message, and I reconnect to the WS, but it seems not to be working correctly. 

Can you please confirm the timing of non activity of websocket (10-20-30 min ? ), is there anyway I can debug the activity of WS, and finally anayway I could properly handle that ? 

In advance thanks for the reply.

**Comment by Steejayy** [comment]
Additionnaly, I sent a ping event every five minutes to the server, but still the same error. 

Thanks in advance for the reactivity 💯

**Comment by f3rno** [comment]
@Steejayy can you provide a stack trace/specific error message? This is not enough info to debug

**Comment by Steejayy** [comment]
hello @f3rno  sorry for the late reply. 
------------------------------------------------------
This is the stack : 
 Error: ws not open
    at WSv2.send (/home/ubuntu/dte/source/node_modules/bitfinex-api-node/lib/transports/ws2.js:1135:33)
    at Promise (/home/ubuntu/dte/source/node_modules/bitfinex-api-node/lib/transports/ws2.js:219:12)
    at Promise._execute (/home/ubuntu/dte/source/node_modules/bluebird/js/release/debuggability.js:303:9)
    at Promise._resolveFromExecutor (/home/ubuntu/dte/source/node_modules/bluebird/js/release/promise.js:483:18)
    at new Promise (/home/ubuntu/dte/source/node_modules/bluebird/js/release/promise.js:79:10)
    at WSv2.auth (/home/ubuntu/dte/source/node_modules/bitfinex-api-node/lib/transports/ws2.js:213:12)
    at WSv2.XFinexOrdersMonitor.ws.on (/home/ubuntu/dte/source/dist/models/exchanges/xfinex/orders-monitor.js:67:16)
    at WSv2.emit (events.js:182:13)
    at WSv2.EventEmitter.emit (domain.js:442:20)
    at WSv2._onWSOpen (/home/ubuntu/dte/source/node_modules/bitfinex-api-node/lib/transports/ws2.js:353:10)
    at WebSocket.emit (events.js:187:15)
    at WebSocket.EventEmitter.emit (domain.js:442:20)
    at WebSocket.setSocket (/home/ubuntu/dte/source/node_modules/ws/lib/WebSocket.js:167:10)
    at ClientRequest._req.on (/home/ubuntu/dte/source/node_modules/ws/lib/WebSocket.js:715:10)
    at ClientRequest.emit (events.js:182:13)
    at ClientRequest.EventEmitter.emit (domain.js:442:20)
    at TLSSocket.socketOnData (_http_client.js:465:11)
    at TLSSocket.emit (events.js:182:13)
    at TLSSocket.EventEmitter.emit (domain.js:442:20)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at TLSSocket.Readable.push (_stream_readable.js:219:10)
    at TLSWrap.onread (net.js:639:20)
----------------------------------------------------------------------------

Basically, I would like my websocket connection to be always active so I can open and close orders. 

ws.on("error", async err => {
    logger.error(`${exchangeName}: ${err.stack || err.toString()}`)

    try {
      if (ws.isOpen()) {
        await ws.close()
      }
    } catch (error) {
      logger.error(error)
    }
  })

  ws.on("close", async () => {
    logger.info("ws connection closed, reconnecting..")
    try {
      if (!ws.isOpen()) {
        await ws.reconnect()
        logger.info(`websocket reconnected with status ${ws.isOpen()}`)
      }
    } catch (error) {
      logger.verbose("ws error occured during reconnection")
      logger.error(error)
    }
  })

**Comment by Steejayy** [comment]
Sorry not to flood, but trying to give you all the informations I have. 

When I try reconnecting, I have the error: 
 at ClientRequest._req.on (/home/ubuntu/dte/source/node_modules/ws/lib/WebSocket.js:653:21)
    at ClientRequest.emit (events.js:182:13)
    at ClientRequest.EventEmitter.emit (domain.js:442:20)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:546:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
    at TLSSocket.socketOnData (_http_client.js:432:20)
    at TLSSocket.emit (events.js:182:13)
    at TLSSocket.EventEmitter.emit (domain.js:442:20)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at TLSSocket.Readable.push (_stream_readable.js:219:10)
    at TLSWrap.onread (net.js:639:20)

**Comment by f3rno** [comment]
@Steejayy if you have `autoReconnect` set, then you should not reconnect manually in `on('close', ..` since it already occurs automatically after a delay, after emitting the 'close' event. So your code above will reconnect twice, potentially causing problems.

You should not reconnect on error, but leave `autoReconnect` on and maybe use the packet watchdog to cycle the connection if packets stop arriving. Let me know if this solves your problem

**Comment by Steejayy** [comment]
@f3rno thanks. The problem was that with the autoreconnect I wasnt authenticated. For the moment is working with the manual reconnect. Will let you know if any other issue. thanks for the follow up
